cpp/src/arrow/compute/kernels/vector_selection_benchmark.cc

```diff
@@ -0,0 +1,291 @@
+                                        (1 - args.null_proportion));
+    }
+    auto values = std::static_pointer_cast<StringArray>(rand.String(
+        array_size, string_min_length, string_max_length, args.null_proportion));
+    Bench(values);
+  }
+
+  void Bench(const std::shared_ptr<Array>& values) {
+    bool indices_null_proportion = indices_have_nulls ? args.null_proportion : 0;
+    auto indices =
+        rand.Int32(static_cast<int32_t>(values->length()), 0,
+                   static_cast<int32_t>(values->length() - 1), indices_null_proportion);
+
+    if (monotonic_indices) {
+      auto arg_sorter = *SortToIndices(*indices);
+      indices = *Take(*indices, *arg_sorter);
+    }
+
+    for (auto _ : state) {
+      ABORT_NOT_OK(Take(values, indices).status());
+    }
+  }
+};
+
+struct FilterBenchmark {
+  benchmark::State& state;
+  FilterArgs args;
+  random::RandomArrayGenerator rand;
+  bool filter_has_nulls;
+
+  FilterBenchmark(benchmark::State& state, bool filter_has_nulls)
+      : state(state),
+        args(state, filter_has_nulls),
+        rand(kSeed),
+        filter_has_nulls(filter_has_nulls) {}
+
+  void Int64() {
+    const int64_t array_size = args.size / sizeof(int64_t);
+    auto values = std::static_pointer_cast<NumericArray<Int64Type>>(
+        rand.Int64(array_size, -100, 100, args.values_null_proportion));
+    Bench(values);
+  }
+
+  void FSLInt64() {
+    const int64_t array_size = args.size / sizeof(int64_t);
+    auto int_array = std::static_pointer_cast<NumericArray<Int64Type>>(
+        rand.Int64(array_size, -100, 100, args.values_null_proportion));
+    auto values = std::make_shared<FixedSizeListArray>(
+        fixed_size_list(int64(), 1), array_size, int_array, int_array->null_bitmap(),
+        int_array->null_count());
+    Bench(values);
+  }
+
+  void String() {
+    int32_t string_min_length = 0, string_max_length = 32;
+    int32_t string_mean_length = (string_max_length + string_min_length) / 2;
+    // for an array of 50% null strings, we need to generate twice as many strings
+    // to ensure that they have an average of args.size total characters
+    int64_t array_size = args.size;
+    if (args.values_null_proportion < 1) {
+      array_size = static_cast<int64_t>(args.size / string_mean_length /
+                                        (1 - args.values_null_proportion));
+    }
+    auto values = std::static_pointer_cast<StringArray>(rand.String(
+        array_size, string_min_length, string_max_length, args.values_null_proportion));
+    Bench(values);
+  }
+
+  void Bench(const std::shared_ptr<Array>& values) {
+    auto filter = rand.Boolean(values->length(), args.selected_proportion,
+                               args.filter_null_proportion);
+    for (auto _ : state) {
+      ABORT_NOT_OK(Filter(values, filter).status());
+    }
+  }
+};
+
+static void FilterInt64FilterNoNulls(benchmark::State& state) {
+  return FilterBenchmark(state, false).Int64();
```

Comment:
You might get compiler warnings for returning in a void function.

Reply:
ah thanks, fixing